src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java

```diff
@@ -2488,15 +2561,130 @@ interface DateTimePrinterParser {
+            } else {
+                fractionalDigits = nanoMaxWidth;
+            }
+        }
+
+        private static List<DateTimePrinterParser> buildPrinterParsers(
+                boolean optional,
+                int nanoMinWidth,
+                int nanoMaxWidth
+        ) {
+            List<DateTimePrinterParser> printerParsers = new ArrayList<>();
+            printerParsers.add(new NumberPrinterParser(ChronoField.HOUR_OF_DAY, 2, 2, SignStyle.NOT_NEGATIVE));
+            printerParsers.add(new CharLiteralPrinterParser(':'));
+            printerParsers.add(new NumberPrinterParser(ChronoField.MINUTE_OF_HOUR, 2, 2, SignStyle.NOT_NEGATIVE));
+            printerParsers.add(new CharLiteralPrinterParser(':'));
+            printerParsers.add(new NumberPrinterParser(ChronoField.SECOND_OF_MINUTE, 2, 2, SignStyle.NOT_NEGATIVE));
+
+            if (nanoMinWidth >= 0) {
+                DateTimePrinterParser nanosPrinterParser
+                        = new NanosPrinterParser(nanoMinWidth, nanoMaxWidth, optional, 0);
+                if (optional) {
+                    nanosPrinterParser
+                            = new CompositePrinterParser(new DateTimePrinterParser[]{nanosPrinterParser}, true);
+                }
+                printerParsers.add(nanosPrinterParser);
+            }
+
+            return printerParsers;
+        }
+
+        @Override
+        public boolean format(DateTimePrintContext context, StringBuilder buf) {
+            TemporalAccessor temporal = context.getTemporal();
+
+            LocalTime time;
+            if (temporal instanceof LocalTime) {
+                time = (LocalTime) temporal;
+            } else if (temporal instanceof LocalDateTime) {
+                time = ((LocalDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof ZonedDateTime) {
+                time = ((ZonedDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetDateTime) {
+                time = ((OffsetDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetTime) {
+                time = ((OffsetTime) temporal).toLocalTime();
+            } else {
+                return super.format(context, buf);
+            }
+
+            formatTime(buf, time);
+
+            int nano = time.getNano();
+            if (nano != 0 || !nanoOptional) {
+                if (fractionalDigits < 0) {
+                    formatNano(buf, nano);
+                } else {
+                    formatNano(buf, fractionalDigits, nano);
+                }
+            }
+
+            buf.append('Z');
+            return true;
+        }
+    }
+
     //-----------------------------------------------------------------------
     /**
      * Composite printer and parser.
      */
-    static final class CompositePrinterParser implements DateTimePrinterParser {
+    static class CompositePrinterParser implements DateTimePrinterParser {
```

Comment:
This class can be `sealed`:
```suggestion
    static sealed class CompositePrinterParser implements DateTimePrinterParser {
```

Reply:
Changes have been made based on your suggestions